### DISCUSSION TITLE
[help] tracking dependency of manually constructed targets in `tar_read`

### DISCUSSION THREAD

**Discussion by shirdekel** [ropensci/targets#1150] [Help]
### Help

- [x] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'm creating automated reports using a report using a template file and `knit_expand`. However, this involves constructing target names manually to call within `tar_read`, which means that there's no dependency. Is there a better way of doing this?

`report.qmd`:

````qmd
---
title: 'report.qmd source file with parameters'
output_format: html_document
params:
  values: 0
---

assume these lines are in report.qmd.

```{r}
#| label: child
#| results: asis
src <-
  params$values |>
  purrr::pmap(
    function(x, y) {
      knitr::knit_expand(here::here("template.qmd"), x = x, y = y)
    }
  ) |>
  paste(collapse = "\n")
knitr::knit_child(text = src, quiet = TRUE) |>
  cat()
```

````

`template.qmd`:

````qmd
---
title: 'template file'
---
assume these lines are in template.qmd.

```{r}
print(targets::tar_read('test_{{x}}_{{y}}'))
```

````

`_targets.R`:

```r
library(targets)
library(tarchetypes)
values <- tibble::tribble(
  ~x, ~y,
  "a1", "b1",
  "a2", "b2"
) |>
  dplyr::mutate(names = paste(x, y, sep = "_"))
list(
  tar_map(
    values = values,
    names = names,
    tar_target(
      test,
      paste(x, y)
    )
  ),
  tar_quarto(
    report,
    path = "report.qmd",
    execute_params = list(
      values = values |>
        dplyr::select(x, y)
    ),
    extra_files = "template.qmd"
  )
)
```

**Comment by wlandau** [comment]
This use case is highly custom, and `tar_quarto()` is not designed for it. I recommend creating your own custom target which declares upstream dependencies, writes the qmd source, and renders it.

```r
tar_target(
 name =  report,
 command = {
    dependency_target
    other_dependency_target
    write_qmd_source_file()
    quarto::quarto_render("qmd_source_file.qmd")
    c("qmd_source_file.qmd", "qmd_source_file.html")
  },
  format = "file"
)
```

**Reply by shirdekel** [reply]
Thanks!

Is writing the `qmd` source necessary for the dependency? Would `targets` not track it by virtue of it being declared in the final character vector?

**Reply by shirdekel** [reply]
And is there an automated way of including targets from a list generated by `tar_map` (named `target_map` here)? I tried the below, but the target is constantly invalidated and in `tar_visnetwork` the dependency is represented by `target_map` and not connected to the actual targets.

```r
target_map[[2]][tidyselect::eval_select(
          everything(),
          target_map[[2]]
        )]
```

I also tried recreating the relevant target names and wrapping them in `syms()`, which didn't work either. Only manually adding in the targets works.

**Reply by wlandau** [reply]
> Is writing the qmd source necessary for the dependency? Would targets not track it by virtue of it being declared in the final character vector?

Only if your use case requires it. If it exists beforehand, you could us it as is.

> is there an automated way of including targets from a list generated by tar_map (named target_map here)?
> I also tried recreating the relevant target names and wrapping them in syms(), which didn't work either. Only manually adding in the targets works.

Metaprogramming can substitute symbols into expressions, c.f. https://wlandau.github.io/targetopia/contributing.html#metaprogramming and https://books.ropensci.org/targets/static.html#metaprogramming. It can be tricky to get right the first time, and `tar_manifest()` and `tar_glimpse()` are good checks.

**Reply by shirdekel** [reply]
I managed to get a dependency working with `tar_target_raw` and putting `sym("target")` in `command`, but adding more than one gets the error `the command of target example_target must have length 1.`. `tar_eval` isn't useful to me here because I don't need to loop. I just want to link a dependency between the quarto target and a number of upstream targets that I refer to within it.

**Reply by wlandau** [reply]
Maybe something like this?

```r
library(rlang)
library(tarchetypes)
target <- tar_eval(
  tar_target(
    name = report,
    command = {
      dependencies
      run_report()
    }
  ),
  values = list(
    dependencies = as.expression(
      expr(list(!!!syms(c("dependency1", "dependency2"))))
    )
  )
)

target[[1]]$command$expr
#> expression({
#>     list(dependency1, dependency2)
#>     run_report()
#> })
```

**Reply by shirdekel** [reply]
Thank you that works!
